core/sr-io/src/lib.rs

```diff
@@ -101,40 +102,57 @@ export_api! {
 		fn storage(key: &[u8]) -> Option<Vec<u8>>;
 
 		/// Get `key` from child storage and return a `Vec`, empty if there's a problem.
-		fn child_storage(storage_key: &[u8], key: &[u8]) -> Option<Vec<u8>>;
+		fn child_storage(child_trie: ChildTrieReadRef, key: &[u8]) -> Option<Vec<u8>>;
 
 		/// Get `key` from storage, placing the value into `value_out` (as much of it as possible) and return
 		/// the number of bytes that the entry in storage had beyond the offset or None if the storage entry
 		/// doesn't exist at all. Note that if the buffer is smaller than the storage entry length, the returned
 		/// number of bytes is not equal to the number of bytes written to the `value_out`.
 		fn read_storage(key: &[u8], value_out: &mut [u8], value_offset: usize) -> Option<usize>;
 
+		/// Get child trie for a given `storage_key` location, or `None` if undefined.
+		fn child_trie(storage_key: &[u8]) -> Option<ChildTrie>;
+
+		/// Update or create an existing child trie.
+		/// Return false if it could not be updated (eg direct change
+		/// of root is not allowed).
+		/// This is accessible to runtime module, but is not safe for
+		/// direct access, (eg contract).
+		/// Problematic use case is direct construction of child trie with existing root or keyspace.
+		/// `ChildTrie` input need to be fetch or instantiate from a valid `Keyspace` generator.
+		fn set_child_trie(ct: ChildTrie) -> bool;
```

Comment:
Can this be replaced with a safer semantics? Like `init_child_trie(storage_key: &[u8])` which returns a new unique keyspace (instead of it being passed in) and rejects if the parent trie is not empty?

Reply:
Isn't it the same thing as what we do with `fetch_or_new` method?
I need at some point a method to update child trie extension content, this was reducing the number of methods over the wasm boundary, but yes it can be split between creation and update case.

Reply:
The main suggestion here is letting this call generate the keyspace for newly initialized child tries.

The function could just then be `set_child_trie(storage_key: &[u8], extension: &[u8]) -> ChildTrie`, which fetches or initializes a child trie at the storage key (parent path) and sets its extension data. It might be able to replace `fetch_or_new` entirely.